Composer/packages/client/src/pages/botProject/table-view.tsx

```diff
@@ -0,0 +1,736 @@
+  );
+
+  const updatePublishTarget = useCallback(
+    async (name: string, type: string, configuration: string) => {
+      if (!editTarget) {
+        return;
+      }
+
+      const targets = publishTargets ? [...publishTargets] : [];
+
+      targets[editTarget.index] = {
+        name,
+        type,
+        configuration,
+      };
+
+      await setPublishTargets(targets, projectId);
+    },
+    [publishTargets, projectId, botName, editTarget]
+  );
+
+  useEffect(() => {
+    setDialogProps({
+      title: formatMessage('Add a publish profile'),
+      type: DialogType.normal,
+      children: (
+        <CreatePublishTarget
+          closeDialog={() => setAddDialogHidden(true)}
+          current={null}
+          targets={publishTargets || []}
+          types={publishTypes}
+          updateSettings={savePublishTarget}
+        />
+      ),
+    });
+  }, [publishTypes, savePublishTarget, publishTargets]);
+
+  useEffect(() => {
+    setEditDialogProps({
+      title: formatMessage('Edit a publish profile'),
+      type: DialogType.normal,
+      children: (
+        <CreatePublishTarget
+          closeDialog={() => setEditDialogHidden(true)}
+          current={editTarget ? editTarget.item : null}
+          targets={(publishTargets || []).filter((item) => editTarget && item.name != editTarget.item.name)}
+          types={publishTypes}
+          updateSettings={updatePublishTarget}
+        />
+      ),
+    });
+  }, [editTarget, publishTypes, updatePublishTarget]);
+
+  const onAddLangModalSubmit = async (formData) => {
+    await addLanguages({
+      ...formData,
+      projectId,
+    });
+  };
+
+  const languageListOptions = useMemo(() => {
+    const languageList = languageListTemplates(languages, locale, defaultLanguage);
+    const enableLanguages = languageList.filter(({ isEnabled }) => !!isEnabled);
+    return enableLanguages.map((item) => {
+      const { language, locale } = item;
+      return {
+        key: locale,
+        title: locale,
+        text: language,
+      };
+    });
+  }, [languages]);
+
+  const onRenderLabel = (props: ITextFieldProps | undefined) => {
+    return (
+      <div css={labelContainer}>
+        <div css={customerLabel}> {props?.label} </div>
+        <TooltipHost content={props?.label}>
+          <Icon iconName={'Unknown'} styles={unknownIconStyle(props?.required)} />
+        </TooltipHost>
+      </div>
+    );
+  };
+
+  const SkillHostEndPoint = useMemo(() => {
+    return (
+      <TextField
+        aria-labelledby={'SkillHostEndPoint'}
+        label={formatMessage('Skill host endpoint url')}
+        placeholder={'Enter Skill host endpoint url'}
+        value={skillHostEndpoint}
+        onChange={async (e, value) =>
+          await setSettings(projectId, {
+            ...settings,
+            skillHostEndpoint: value,
+          })
+        }
+        onRenderLabel={onRenderLabel}
+      />
+    );
+  }, [projectId]);
+
+  const AppIdOrPassWord = useMemo(() => {
```

Comment:
Maybe make this it's own component instead of using useMemo?
you can use React.memo on the new component to prevent excessive renders